runtime/lua/vim/lsp/buf.lua

```diff
@@ -116,31 +116,32 @@ end
 --- asks the user to select one.
 --
 ---@returns The client that the user selected or nil
-local function select_client(method)
-  local clients = vim.tbl_values(vim.lsp.buf_get_clients());
-  clients = vim.tbl_filter(function (client)
+local function select_client(method, on_choice)
+  validate {
+    on_choice = { on_choice, 'function', false },
+  }
+  local clients = vim.tbl_values(vim.lsp.buf_get_clients())
+  clients = vim.tbl_filter(function(client)
     return client.supports_method(method)
   end, clients)
   -- better UX when choices are always in the same order (between restarts)
-  table.sort(clients, function (a, b) return a.name < b.name end)
+  table.sort(clients, function(a, b)
+    return a.name < b.name
+  end)
 
   if #clients > 1 then
-    local choices = {}
-    for k,v in pairs(clients) do
-      table.insert(choices, string.format("%d %s", k, v.name))
-    end
-    local user_choice = vim.fn.confirm(
-      "Select a language server:",
-      table.concat(choices, "\n"),
-      0,
-      "Question"
-    )
-    if user_choice == 0 then return nil end
-    return clients[user_choice]
+    vim.ui.select(clients, {
+      prompt = 'Select a language server:',
+      format_item = function(client)
+        return client.name
+      end,
+    }, function(item, idx)
+      on_choice(item)
+    end)
   elseif #clients < 1 then
-    return nil
+    on_choice(0)
```

Comment:
```suggestion
    on_choice(nil)
```

From the docs:

```
                    {on_choice}  function ((item|nil, idx|nil) -> ()) Called
                                 once the user made a choice. `idx` is the
                                 1-based index of `item` within `item` . `nil`
                                 if the user aborted the dialog.
```

It shouldn't be called with `0` if there are no items or if the user aborted

Reply:
I think that was a leftover from when I was prototyping, I initially thought the return value was the index